File: TypeDefinitions/GameObjectProxy.cs
namespace WorldServer.Scripting.MoonSharp.Proxies;

using global::MoonSharp.Interpreter;
using JetBrains.Annotations;
using World.Objects;

[PublicAPI]
public class GameObjectProxy : UnitProxy
{
    private readonly GameObject _gameObject;

    [MoonSharpHidden]
    public GameObjectProxy(GameObject gameObject)
        : base(gameObject)
    {
        _gameObject = gameObject;
    }

    [PublicAPI]
    public uint Entry => _gameObject.Entry;

    [PublicAPI]
    public bool Interactable => _gameObject.Interactable;

    [PublicAPI]
    public void SetState(string stateName, int state) => _gameObject.ScriptState.SetState(stateName, state);

    [PublicAPI]
    public int? GetState(string stateName) => _gameObject.ScriptState.GetState(stateName);

    [PublicAPI]
    public byte VfxState { get => _gameObject.VfxState; set => _gameObject.VfxState = value; }

    [PublicAPI]
    public bool NoRespawn { get => _gameObject.NoRespawn; set => _gameObject.NoRespawn = value; }
}

File: TypeDefinitions/GameObjectProxy.lua
-- game_object_proxy.lua

---@class GameObjectProxy : UnitProxy
---@field Entry number
---@field Interactable boolean
---@field VfxState number
local GameObjectProxy = {}

--- Sets the state of the game object.
---@param stateName string
---@param state number
function GameObjectProxy:SetState(stateName, state) end

--- Gets the state of the game object.
---@param stateName string
---@return number|nil
function GameObjectProxy:GetState(stateName) end

return GameObjectProxy
